app/lib/dartdoc/dartdoc_runner.dart

```diff
@@ -19,7 +28,68 @@ class DartdocRunner implements TaskRunner {
 
   @override
   Future<bool> runTask(Task task) async {
-    // TODO: implement doc generation and upload
-    return false; // no race detected
+    final tempDir =
+        await Directory.systemTemp.createTemp('pub-dartlang-dartdoc');
+    final pubCacheDir = await tempDir.resolveSymbolicLinks();
+
+    final pubEnv = new PubEnvironment(pubCacheDir: pubCacheDir);
+
+    try {
+      final pkgLocation =
+          await pubEnv.getLocation(task.package, version: task.version);
+      final pkgPath = pkgLocation.location;
+      await _runDartdoc(task, pkgPath);
+      await _writeMetadata(task, pkgPath);
```

Comment:
If `_runDartdoc` failed, this call will fail, since the `doc` folder is not there.

Reply:
`TaskScheduler` has a proper error handling, so we don't really need to worry about handling each of these errors separately. In fact, not having a `doc` directory helps us to get an exception without explicitly checking for it (and then throwing ourselves).

Reply:
It's generally not advisable running commands / steps if they have assumptions about previous commands without ensuring that these previous commands completed successfully.

That's precisely why shell scripts often use `set -e` to abort after the first failed command.

Just assume that the package author already had a `doc` folder, then we might accidentally upload that, incorrect doc folder, to cloud storage and claim everything is fine, even though we didn't generate proper docs.

I'll add a comment further down about it.